addons/loyalty/models/loyalty_program.py

```diff
@@ -445,6 +445,29 @@ def toggle_active(self):
         return res
 
     def write(self, vals):
+        # Handle the case of unarchiving a program
+        if vals.get('active'):
+            all_codes = [
+                rule.code for rule in self.with_context(active_test=False).rule_ids if rule.code
+            ]
+            # Find active rules that conflict with these codes
+            conflicting_codes = self.env['loyalty.rule'].search_fetch(
+                [('code', 'in', all_codes), ('program_id.active', '=', True)],
+                ['code']
+            ).mapped('code')
+            # Detect duplicate codes within this batch
+            duplicates_in_batch = [c for c, count in Counter(all_codes).items() if count > 1]
+            if conflicting_codes:
+                raise ValidationError(_(
+                    'Cannot unarchive programs because active programs with the same code already'
+                    ' exist: %s', ', '.join(conflicting_codes),
```

Comment:
```suggestion
                    "Cannot unarchive programs because active programs with the same code already"
                    " exist: %s", ', '.join(conflicting_codes),
```